gradle-jdks/src/main/java/com/palantir/gradle/jdks/GradleWrapperPatcher.java

```diff
@@ -0,0 +1,182 @@
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.inject.Inject;
+import org.apache.commons.io.IOUtils;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.ProjectLayout;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+@AutoParallelizable
+public abstract class GradleWrapperPatcher {
+
+    private static final Logger log = Logging.getLogger(GradleWrapperPatcher.class);
+    private static final String ENABLE_GRADLE_JDK_SETUP = "gradle.jdk.setup.enabled";
+    private static final String GRADLEW_PATCH = "gradlew-patch";
+    private static final String GRADLEW_UNIX_SCRIPT = "gradlew";
+    private static final String COMMENT_BLOCK = "###";
+    private static final String SHEBANG = "#!";
+
+    // DO NOT CHANGE the header and the footer, they are used to identify the patch block
+    private static final String GRADLEW_PATCH_HEADER = "# >>> Gradle JDK setup >>>";
+    private static final String GRADLEW_PATCH_FOOTER = "# <<< Gradle JDK setup <<<";
+
+    interface Params {
+        @Inject
+        ProjectLayout getProjectLayout();
+    }
+
+    public abstract static class GradleWrapperPatcherTask extends GradleWrapperPatcherTaskImpl {}
+
+    static void action(Params params) {
+        if (!isGradleJdkSetupEnabled(params.getProjectLayout())) {
+            return;
+        }
+        log.lifecycle("Gradle JDK setup is enabled, patching the gradle wrapper files");
+        Path gradlewFile = params.getProjectLayout()
+                .files(GRADLEW_UNIX_SCRIPT)
+                .getSingleFile()
+                .toPath();
+        patchGradlewContent(gradlewFile);
+    }
+
+    private static Properties loadGradleProperties(File gradleWrapperPropsFile) {
+        Properties properties = new Properties();
+        try (FileInputStream inputStream = new FileInputStream(gradleWrapperPropsFile)) {
+            properties.load(inputStream);
+            return properties;
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("Unable to find gradle.properties file", e);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not read gradle.properties file", e);
+        }
+    }
+
+    private static void write(Path destPath, String content) {
+        try {
+            Files.write(destPath, content.getBytes(StandardCharsets.UTF_8));
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to write file", e);
+        }
+    }
+
+    private static boolean isGradleJdkSetupEnabled(ProjectLayout projectLayout) {
+        FileCollection gradleProperties = projectLayout.files("gradle.properties");
```

Comment:
Rather than manually reading the `gradle.properties`, we should just use `Project.property(...)` to find it so we get [the hierarchy support](https://docs.gradle.org/8.7/userguide/build_environment.html).

The trick to do this in an autoparallelisable-friendly way is to add a param called `getGradleJdksEnabled()` and set it's default value to be something that looks up the property using the Project object.

Reply:
Actually, I wonder if we could just use `onlyIf` here instead